doc/bulk_counter/bulk_counter.md

```diff
@@ -0,0 +1,145 @@
+  - SAI_STATS_MODE_READ -> SAI_STATS_MODE_BULK_READ
+  - SAI_STATS_MODE_READ_AND_CLEAR -> SAI_STATS_MODE_BULK_READ_AND_CLEAR
+
+### Architecture Design
+
+For each counter group, different statistic type is allowed to chooose bulk or non-bulk API based on SAI capability.
+
+![architecture](/doc/bulk_counter/bulk_counter.svg).
+
+> Note: In the picture, pg/queue watermark statistic use bulk API and buffer watermark statistic uses non-bulk API. This is just an example to show the design idea. 
+
+### High-Level Design
+
+Changes shall be made to sonic-sairedis to support this feature. No CLI change. No DB schema change.
+
+> Note: Code present in this design document is only for demonstrating the design idea, it is not production code.
+
+#### sonic-sairedis
+
+##### Bulk Statistic Context
+
+A new structure shall be added to FlexCounter class. 
+
+This structure is created because:
+
+- Meet the signature of sai_bulk_object_get_stats and sai_bulk_object_clear_stats
+- Avoid constructing these information each time collecting statistic. The bulk context shall only be updated under below cases:
+  - New object join counter group. E.g. adding a new port object.
+  - Existing object leave counter group. E.g removing an existing port object.
+
+```cpp
+struct BulkStatsContext
+{
+    bool support = true;
+    std::vector<sai_object_id_t> object_vids;
+    std::vector<sai_object_key_t> object_keys; 
+    std::vector<sai_stat_id_t> counter_ids;
+    std::vector<sai_status_t> get_statuses;
+    std::vector<sai_status_t> clear_statuses;
+    std::vector<uint64_t> counters;
+    std::shared_ptr<sai_stat_capability_list_t> stats_capas;
+};
+```
+- support: indicate if bulk is supported for this object type.
+- object_keys: objects that participate the bulk call. E.g. for port, SAI object id value shall be put into sai_object_key_t structure.
+- counter_ids: SAI statistic IDs that will be queried/cleared by the bulk call.
+- get_statuses: SAI bulk get return value for each object.
+- clear_statuses: SAI bulk clear return value for each object. Only effective if stats mode is read and clear.
+- counters: counter values that will be fill by vendor SAI.
+- stats_capas: stats capability for each statitstic IDs for current object type.
+
+The flow of how to updating bulk context will be discussed in following section.
+
+Instance of BulkStatsContext shall be added to FlexCounter class for each object type. One object type may have multiple BulkStatsContext instances.
+
+```cpp
+
+BulkStatsContext m_portBulkContext;
+BulkStatsContext m_priorityGroupBulkContext;
+BulkStatsContext m_queueBulkContext;
+...
+
+```
+
+##### Update Bulk Context
+
+1. New object join counter group.
+
+![Add Object Flow](/doc/bulk_counter/object_join_counter_group.svg).
+
```

Comment:
In the figure:

1. Why needs to check 'Stats capability available?' Should always 'Query stats capability' after 'Extract object ID...'
2. If success and 'All counter IDs support bulk', this object is a candidate for bulk get; otherwise bulk not supported.
3. If it is a candidate, check bulkStatsContext list to find the matched one according to stats capabilities.
4. If found, add it to the matched bulkStatsContext
5. If not found, add it to a new bulkStatsContext, then add the context to context list

Reply:
```
Why needs to check 'Stats capability available?' Should always 'Query stats capability' after 'Extract object ID' 
```
Stat capability is usually static data stored in SAI, so if we got it once and save it somewhere, we don't need to get it anymore.

For 2~5, I think you suggested to put bulkStatsContext to a list so that we don't need to manage them for each type. That is a great idea. However, it requires a refactor for class FlexCounter. As you can see in FlexCounter, each object type has similar logic:
```
setXXXCounterList
removeXXX
queryXXXSupportedCounters
getXXXSupportedCounters
collectXXXCounters
```
There are many reasons that we need a refactor:
1. The SAI bulk API call for each object is exactly the same, however, the code to save the stats data to COUNTERS DB is not exactly the same for each object.
2. If SAI bulk API call fails, it shall fall back to non-bulk API to keep backward compatible, and those non-bulk API is encapsulated into collectPortCounters, collectQueueCounters...
3. so on

I am working on a proposal to refactor the FlexCounter, it will be ready within this week.

Reply:
1. IMO, stats capability is per object query, should not be cached in syncd
2. Even each type may have a bulkStatsContext list